chore: Remove embedded Python vllm and sglang engines

grahamking opened the pull request:
What a cleanup! A breath of fresh air.

vllm and sglang are now the sub-process engines from #954

This means unless you build with `--feature python`, dynamo-run does not link `libpython`.

grahamking committed:
chore: Remove embedded Python vllm and sglang engines

What a cleanup! A breath of fresh air.

vllm and sglang are now the sub-process engines from #954

This means unless you build with `--feature python`, dynamo-run does not
link `libpython`.

rmccorm4 commented on container/Dockerfile.none:
```diff
@@ -46,7 +46,7 @@ ARG CARGO_BUILD_JOBS
 
 ENV CARGO_TARGET_DIR=/workspace/target
 
-RUN cargo build --release --locked --features mistralrs,sglang,vllm,python && \
```
no mistral?

rmccorm4 replied:
I'm guessing it's a default feature, but just double checking

grahamking replied:
```
[features]
default = ["mistralrs", "llamacpp"]
```
Oh yeah!

I figured out that llamacpp gets statically linked, so we can include it by default effectively for free.

On top of that vllm and sglang are now sub-process, so they don't need a build flag. My next task is removing the build-your-own-engine ("python" feature flag). That will leave us with only cuda/metal/vulkan as feature flags.

grahamking replied:
I put the mistralrs flag back, might as well be explicit. I also removed llamacpp from default features as that should be it's own commit.

rmccorm4 replied:
Sounds good to me - thanks 🚀

rmccorm4 commented on container/Dockerfile.tensorrt_llm:
```diff
@@ -173,7 +173,7 @@ COPY launch /workspace/launch
 
 # Build Rust crate binaries packaged with the wheel
-RUN cargo build --release --locked --features mistralrs,sglang,vllm,python \
+RUN cargo build --release --locked --features python \
```
no mistral?

rmccorm4 commented on container/Dockerfile.vllm:
```diff
@@ -305,7 +307,7 @@ COPY launch /workspace/launch
 
 # Build Rust crate binaries packaged with the wheel
-RUN cargo build --release --locked --features mistralrs,sglang,vllm,python \
+RUN cargo build --release --locked --features python \
```
no mistral?

rmccorm4 commented on docs/guides/dynamo_run.md:
```diff
@@ -4,8 +4,9 @@
     * [Setup](#setup)
+    * [MistralRS](#mistralrs)
     * [Sglang](#sglang)
     * [lama.cpp](#llama_cpp)
```
just noticed typo 
```suggestion
    * [llama.cpp](#llama_cpp)
```

grahamking replied:
Done, thanks.

rmccorm4 commented on docs/guides/dynamo_run.md:
```diff
@@ -125,15 +133,14 @@ source $HOME/.cargo/env
 #### Step 3: Build
 
-Run `cargo build` to install the `dynamo-run` binary in `target/debug`.
+Run `cargo build --features cuda -p dynamo-run` to install the `dynamo-run` binary in `target/debug`.
```
I think this step is intended to be runnable on macos, and there's a specific shout out for building with cuda support on linux+gpu below -- do we need `--features cuda` in this spot?

rmccorm4 commented on docs/guides/dynamo_run.md:
```diff
@@ -155,57 +162,52 @@ cd target/debug
+2. Run
 
-### llama_cpp
+Any example above using `out=sglang` will work, but our sglang backend is also multi-gpu.
```
```suggestion
Any example above using `out=sglang` will work, but our sglang backend also supports multi-gpu through extra args such as `--tensor-parallel-size`:
```

grahamking replied:
I need to re-test, so will leave out for now and update when re-tested.

grahamking replied:
Got a cluster, tested both vllm and sglang new (subprocess) engines, it works. Added docs to the guide on how to do multi-node in both vllm and sglang.

It's much simpler than before. We're effectively plugging a normal vllm or sglang build into our networking and pre-processing setup. The rest we don't interfere with.

rmccorm4 replied:
Nice!

rmccorm4 commented on docs/guides/dynamo_run.md:
```diff
@@ -155,57 +162,52 @@ cd target/debug
 ### vllm
 
-Using the [vllm](https://github.com/vllm-project/vllm) Python library. We only use the back half of vllm, talking to it over `zmq`. Slow startup, fast inference. Supports both safetensors from HF and GGUF files.
+Using the [vllm](https://github.com/vllm-project/vllm) Python library. Slow startup, fast inference. Supports both safetensors from HF and GGUF files, but is very slow for GGUF - prefer llamacpp.
```
Don't want to upset anyone by saying vllm has a slow startup 😅 
```suggestion
Using the [vllm](https://github.com/vllm-project/vllm) Python library. Supports both safetensors from HF and GGUF files, but is very slow for GGUF - prefer llamacpp.
```

grahamking replied:
It really does though. It's painful compared to mistralrs or llamacpp.

rmccorm4 commented on docs/guides/dynamo_run.md:
```diff
@@ -447,7 +442,7 @@ The output looks like this:
 ### Defaults
 
-The input defaults to `in=text`. The output will default to `mistralrs` engine. If not available whatever engine you have compiled in (so depending on `--features`).
+The input defaults to `in=text`. The output will default to `mistralrs` engine, unless it is disabled with `--no-default-features` in which case vllm is used.
```
`out=mistralrs` for symmetry with `in=text`
```suggestion
The input defaults to `in=text`. The output will default to `out=mistralrs` engine, unless it is disabled with `--no-default-features` in which case vllm is used.
```

rmccorm4 replied:
Also this answered my above comments on removing `--features mistralrs` from docker builds

rmccorm4 commented on launch/dynamo-run/Cargo.toml:
```diff
@@ -68,15 +62,3 @@ clap = { version = "4.5", features = ["derive", "env"] }
 futures-util = { version = "0.3" }
 regex = "1"
-
-[target.x86_64-unknown-linux-gnu.dependencies]
-netlink-packet-route = { version = "0.19", optional = true }
-rtnetlink = { version = "0.14", optional = true }
-
-[target.x86_64-unknown-linux-musl.dependencies]
-netlink-packet-route = { version = "0.19", optional = true }
-rtnetlink = { version = "0.14", optional = true }
-
-[target.aarch64-unknown-linux-gnu.dependencies]
-netlink-packet-route = { version = "0.19", optional = true }
-rtnetlink = { version = "0.14", optional = true }
```
w00t, right after I needed to add `aarch64` sections too 😂

rmccorm4 commented on docs/guides/dynamo_run.md:
```diff
@@ -155,57 +162,52 @@ cd target/debug
+2. Run
 
-### llama_cpp
+Any example above using `out=sglang` will work, but our sglang backend is also multi-gpu.
```
Does multi-node support still work well?

grahamking replied:
I don't know. I was able to test tensor parallel on an 8xL40S box yesterday. Today I'll try to find a second one.

grahamking replied:
Tested, multi-node works well.

rmccorm4 reviewed (approved):
LGTM other than minor suggestions - please address them before merge

grahamking committed:
chore: Restore and update

- Restore listing all used features in build command for clarity.
- Remove llamacpp from default features, that will be it's own commit
  (need to address CI libclang).
- Doc fixes from feedback.

grahamking committed:
chore: Replace bring-your-own-engine with write-your-own-engine

Instead of passing a Python file to `dynamo-run` you can now use
`dynamo-llm` bindings in a regular Python script. Much simpler.

Update the docs, and remove the `pytok` engine. We keep the `pystr`
engine for now because trt-llm uses it.
